### PR TITLE
Set up circleci to publish to gemfury

### PR DESCRIPTION
**Pull request by phamtuan2993**
@nguyenquangminh0711 please check and confirm if this meets your requirements.

# Related task (URL):
https://docs.google.com/spreadsheets/d/1yQhywvqRZk4lifuK_SCIYTBE4L2PL5boqgUFjudJrmw/edit#gid=0&range=45:45 (this link to the current range of the task on the sheet, if changed, please search for _"Migrate internal gems into gemfury"_, thanks)

# Summary
As per conversation, now we would need to update the circle ci process on our eh gems to able to build and push the built gem to gemfury.

> Currently, we have some internal gems used in the Gemfile. Most of them are identified with Git repo and branch. Therefore, we got a problem with versioning, performance and deployment. I suggest to migrate all those gems into gemfury.

The process now would auto be handled by circle ci, take all the needed info and run the build job at the end of any successfully build on master branch.

# Changes proposed in this pull request:

Now each gem would have a .circleci/config.yml that contains all the needed jobs and steps to run on a circle ci process and call for a scripts/publish.rb at the end of the successfully build.

The scripts/publish.rb would almost the same for all gems as it only get the gem info, execute the  command on circle ci container to build the gem and push to gemfury.

For this able to be done, we now move all the info of gem name and gem version to a app.json that located at the root folder (along with the .gemspec). And whenever a dev want to bumb the version, just go and update it there.

screen shot 2017-12-13 at 9 41 12 am

Also as the publish job would use circle ci env variables to get the gemfury token and package folder, please specify it on the circle ci settings for the project.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+version: 2
+jobs:
+  publish:
+    docker:
+      - image: ehdevops/ruby240:1.0.4
```
Comment (thukim) [inline .circleci/config.yml]: same as Minh suggested `ruby240:latest`

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+workflows:
+  version: 2
+  build_and_deploy:
+    jobs:
+      - publish:
+          context: org-global
+          filters:
+            branches:
+              only: master
```
Comment (thukim) [inline .circleci/config.yml]: missing new line

File: app.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  "name": "xero_gateway",
+  "version": "3.2.22",
+  "documents": [
+    {
+      "name": "README",
+      "source": "README.textile",
+      "format": "markdown"
+    }
+  ]
+}
```
Comment (thukim) [inline app.json]: missing new line

File: scripts/publish.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'json'
+
+manifest_file = File.read(File.expand_path('../../app.json', __FILE__))
+app = JSON.parse(manifest_file)
+gemspec_file_name = Dir.glob("./*.gemspec")[0].split('/')[-1]
+
+exec("gem build #{gemspec_file_name} && curl -F package=@#{app['name']}-#{app['version']}.gem https://#{ENV['GEMFURY_TOKEN']}@push.fury.io/#{ENV['GEMFURY_PACKAGE']}/")
```
Comment (thukim) [inline scripts/publish.rb]: missing new line

**Comment by thukim** [comment]
`game name and game version` :trollface: probably you wrote this PR desc while playing games.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+version: 2
+jobs:
+  publish:
+    docker:
+      - image: ehdevops/ruby240:latest
+    working_directory: ~/xero_gateway
+    steps:
+      - checkout
+      - run: bundle check || bundle install --path=vendor/bundle --jobs=4 --retry=3
+      - run: bundle exec ruby scripts/publish.rb
```
Comment (thukim) [inline .circleci/config.yml]: Have you added env vars `GEMFURY_PACKAGE` and `GEMFURY_TOKEN` to circleCI yet?

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+workflows:
+  version: 2
```
Comment (thukim) [inline .circleci/config.yml]: what does this workflows version mean?

File: app.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "xero_gateway",
+  "version": "3.2.22",
```
Comment (thukim) [inline app.json]: if we use the version defined in this file from now onwards, why don't you delete the `version` file?

Comment (thukim) [reply app.json]: so that it wont confuse other devs next time if they update this gem.

File: .circleci/config.yml (hunk 6 above)
Comment (phamtuan2993) [reply .circleci/config.yml]: this come from the default settings for 2.0 circleci
from their doc: https://circleci.com/docs/2.0/workflows/
> add a new workflows: section to the end of your existing .circleci/config.yml file with the version and a unique name for the workflow

File: app.json (hunk 7 above)
Comment (phamtuan2993) [reply app.json]: This gem don't have the version file.
I would delete the version file on other gems if exists, like with feature_toggle here https://github.com/Thinkei/feature_toggle/pull/1/files

Thanks

File: .circleci/config.yml (hunk 5 above)
Comment (phamtuan2993) [reply .circleci/config.yml]: I have informed to Minh that I would only take care for the update on circleci process, the other steps include adding github project into circle or providing env vars would be taken care by Minh.

Also I believe that our gems should/would share the same env vars on this.

Thanks

File: .circleci/config.yml (hunk 2 above)
Comment (phamtuan2993) [reply .circleci/config.yml]: All missing new line added

Thanks

File: lib/oauth/oauth_consumer.rb [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
         http.cert = options[:ssl_client_cert]
         http.key  = options[:ssl_client_key]
       end
     end
-  
-    alias_method_chain :http, :ssl_client_certificates
-    
+
+    alias_method :ssl_client_certificates, :http
+    alias_method :http, :ssl_client_certificates
```
Comment (phamtuan2993) [inline lib/oauth/oauth_consumer.rb:11]: https://apidock.com/rails/ActiveSupport/CoreExtensions/Module/alias_method_chain

Comment (phamtuan2993) [reply lib/oauth/oauth_consumer.rb:11]: it was wrong. punish me please

**Pull request merged**